pkg/ingester/ingester_ring.go

```diff
@@ -138,6 +149,30 @@ func (cfg *RingConfig) ToLifecyclerConfig() ring.LifecyclerConfig {
 	lc.ID = cfg.InstanceID
 	lc.ListenPort = cfg.ListenPort
 	lc.EnableInet6 = cfg.EnableIPv6
+	lc.RingTokenGenerator = cfg.customTokenGenerator(logger)
 
 	return lc
 }
+
+// customTokenGenerator returns a token generator, which is an implementation of ring.TokenGenerator,
+// according to this RingConfig's configuration. If "spread-min-tokens" token generation strategy is set,
+// customTokenGenerator tries to build and return an instance of ring.SpreadMinimizingTokenGenerator.
+// If it was impossible, if "random-tokens" or any other unsupported token generation strategy is set,
+// an instance of ring.RandomTokenGenerator is returned.
+func (cfg *RingConfig) customTokenGenerator(logger log.Logger) ring.TokenGenerator {
+	switch cfg.TokenGeneratorStrategy {
+	case spreadMinimizingTokenGeneration:
+		tokenGenerator, err := ring.NewSpreadMinimizingTokenGenerator(cfg.InstanceID, cfg.InstanceZone, cfg.SpreadMinimizingZones, logger)
+		if err != nil {
+			level.Warn(logger).Log("msg", "It was impossible to generate an instance of SpreadMinimizingTokenGenerator. RandomTokenGenerator will be used instead.", "err", err)
```

Comment:
[Nit] By convention in Mimir, we don't capitalize log messages. I've never seen us punctuating them either.
```suggestion
			level.Warn(logger).Log("msg", "it was impossible to generate an instance of SpreadMinimizingTokenGenerator. RandomTokenGenerator will be used instead", "err", err)
```